compiler/rustc_ast/src/expand/autodiff_attrs.rs

```diff
@@ -0,0 +1,270 @@
+            DiffMode::Forward => write!(f, "Forward"),
+            DiffMode::Reverse => write!(f, "Reverse"),
+            DiffMode::ForwardFirst => write!(f, "ForwardFirst"),
+            DiffMode::ReverseFirst => write!(f, "ReverseFirst"),
+        }
+    }
+}
+
+pub fn valid_ret_activity(mode: DiffMode, activity: DiffActivity) -> bool {
+    if activity == DiffActivity::None {
+        // Only valid if primal returns (), but we can't check that here.
+        return true;
+    }
+    match mode {
+        DiffMode::Inactive => false,
+        DiffMode::Source => false,
+        DiffMode::Forward | DiffMode::ForwardFirst => {
+            activity == DiffActivity::Dual
+                || activity == DiffActivity::DualOnly
+                || activity == DiffActivity::Const
+        }
+        DiffMode::Reverse | DiffMode::ReverseFirst => {
+            activity == DiffActivity::Const
+                || activity == DiffActivity::Active
+                || activity == DiffActivity::ActiveOnly
+        }
+    }
+}
+fn is_ptr_or_ref(ty: &Ty) -> bool {
+    match ty.kind {
+        TyKind::Ptr(_) | TyKind::Ref(_, _) => true,
+        _ => false,
+    }
+}
+// TODO We should make this more robust to also
+// accept aliases of f32 and f64
+//fn is_float(ty: &Ty) -> bool {
+//    false
+//}
+pub fn valid_ty_for_activity(ty: &P<Ty>, activity: DiffActivity) -> bool {
+    if is_ptr_or_ref(ty) {
+        return activity == DiffActivity::Dual
+            || activity == DiffActivity::DualOnly
+            || activity == DiffActivity::Duplicated
+            || activity == DiffActivity::DuplicatedOnly
+            || activity == DiffActivity::Const;
+    }
+    true
+    //if is_scalar_ty(&ty) {
+    //    return activity == DiffActivity::Active || activity == DiffActivity::ActiveOnly ||
+    //        activity == DiffActivity::Const;
+    //}
+}
+pub fn valid_input_activity(mode: DiffMode, activity: DiffActivity) -> bool {
+    return match mode {
+        DiffMode::Inactive => false,
+        DiffMode::Source => false,
+        DiffMode::Forward | DiffMode::ForwardFirst => {
+            // These are the only valid cases
+            activity == DiffActivity::Dual
+                || activity == DiffActivity::DualOnly
+                || activity == DiffActivity::Const
+        }
+        DiffMode::Reverse | DiffMode::ReverseFirst => {
+            // These are the only valid cases
+            activity == DiffActivity::Active
+                || activity == DiffActivity::ActiveOnly
+                || activity == DiffActivity::Const
+                || activity == DiffActivity::Duplicated
+                || activity == DiffActivity::DuplicatedOnly
+        }
+    };
+}
+pub fn invalid_input_activities(mode: DiffMode, activity_vec: &[DiffActivity]) -> Option<usize> {
+    for i in 0..activity_vec.len() {
+        if !valid_input_activity(mode, activity_vec[i]) {
+            return Some(i);
+        }
+    }
+    None
+}
+
+#[allow(dead_code)]
```

Comment:
I did not find a good solution to having two configurations. One I used in some places is to have two submodules, and depending on which config we run in I just include everything in the main module from the relevant submodule. Shall I do that here too?

Reply:
As mentioned elsewhere, instead of `#[allow(dead_code)]`, we probably want to use lint-expectations in case the lint gets smarter in the future and no longer fires, e.g. `#[expect(dead_code)]`.

Reply:
There is one set of deps which is always used, one set which is only used when the config is set, nd one set which is only used when the config is not set. If we add expect, it would incorrectly trigger in one of these cases, so I don't think we win much here. The only (Brute force) solution I see here is to have two modules and duplicate things by having dummy implementations, then I can have a parent module from where I import things accordingly based on the config. I do that in the backend, but it adds a lot of extra lines of code to get rid of this allow, I'm not sure if it's worth it.

Reply:
In that case, probably leave the allow, but please do leave a FIXME explaining why the `allow`s are needed.

Reply:
done.